File: InterviewBit/sum-root-to-leaf-numbers.py
# https://www.interviewbit.com/courses/programming/topics/trees/problems/sum-root-to-leaf-numbers/
# 

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Integer:
    def __init__(self, x):
        self.val = x

class Solution:
    # @param A : root node of tree
    # @return an integer
    def sumNumbers(self, A):
        sum = Integer(0)
        self.sumRootToLeaves(A, sum, [])
        return sum.val % 1003
        
    def sumRootToLeaves(self, root, sum, currentPath):
        if not root: return
    
        currentPath.append(root.val)
        
        if not (root.left or root.right):
            sum.val += int(''.join(map(str, currentPath)))
            
        self.sumRootToLeaves(root.left, sum, currentPath)
        self.sumRootToLeaves(root.right, sum, currentPath)
        
        del currentPath[-1]
        

File: LeetCode/sum-root-to-leaf-numbers.java
// https://leetcode.com/problems/sum-root-to-leaf-numbers
// 2 ms

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    class IntWrap {
        int val = 0;
    }
    
    public int sumNumbers(TreeNode root) {
        IntWrap count = new IntWrap();
        sumNumbers(root, new StringBuilder(), count);
        return count.val;
    }
    
    private void sumNumbers(TreeNode root, StringBuilder path, IntWrap count) {
        if (root == null) return;
        path.append(String.valueOf(root.val));
        
        if (root.left == null && root.right == null && path.length()!=0) {
            count.val += Integer.parseInt(path.toString());
        }
        
        sumNumbers(root.left, path, count);
        sumNumbers(root.right, path, count);
        path.deleteCharAt(path.length()-1);
    }
}
